islandora_compound_object.module

```diff
@@ -268,40 +268,100 @@ function islandora_compound_object_access(AbstractObject $object) {
 }
 
 /**
- * Replaces parent object thumbnail with child's thumbnail.
+ * Implements hook_compound_cmodel_islandora_derivative().
  */
-function islandora_compound_object_update_parent_thumbnail($parent) {
+function islandora_compound_object_compoundCModel_islandora_derivative(AbstractObject $object = NULL, $ds_modified_params = array()) {
+  $module_path = drupal_get_path('module', 'islandora_compound_object');
+  $derivatives = array();
+  if (variable_get('islandora_compound_object_tn_deriv_hooks', FALSE)) {
+    $derivatives[] = array(
+      'source_dsid' => NULL,
+      'destination_dsid' => 'TN',
+      'function' => array('islandora_compound_object_create_thumbnail'),
+      'file' => "$module_path/includes/derivatives.inc",
+    );
+  }
+
+  return $derivatives;
+
+}
+
+/**
+ * Implements hook_islandora_derivative().
+ */
+function islandora_compound_object_islandora_derivative(AbstractObject $object) {
+  $module_path = drupal_get_path('module', 'islandora_compound_object');
+  $derivatives = array();
+  // Short circuit if feature is disabled.
+  if (variable_get('islandora_compound_object_tn_deriv_hooks', FALSE)) {
+    $derivatives[] = array(
+      'source_dsid' => 'TN',
+      'destination_dsid' => NULL,
+      'function' => array('islandora_compound_object_create_thumbnail_on_parent'),
+      'file' => "$module_path/includes/derivatives.inc",
+    );
+  }
+  return $derivatives;
+}
+
+/**
+ * Update a parent object's TN by deriving it's first child's TN.
+ *
+ * Note: This method does not enforce the relationship between
+ * parent/child if both are provided. It is left up to the caller
+ * to ensure that objects provided have the appropriate relationship.
+ *
+ * @param AbstractObject $parent
+ *   The compound object.
+ * @param AbstractObject $child
+ *   The child object (optional). Note: Ideally, this would be required.
+ *   Keeping it optional for backwards compatibility.
+ * @param bool $force
+ *   Boolean for forcing this derivative creation.
+ */
+function islandora_compound_object_update_parent_thumbnail(AbstractObject $parent, AbstractObject $child = NULL, $force = TRUE) {
```

Comment:
My thought was that if anyone ever needed to call into this method in the future the force logic would be preserved.